### PR TITLE
Support loading of CSI from URLs/streams.

### PR DESCRIPTION
**Pull request by cmnbroad**
Fixes https://github.com/samtools/htsjdk/issues/1507.

### CODE REVIEW

File: src/main/java/htsjdk/samtools/BAMFileReader.java [hunk 1]
```diff
@@ -407,20 +407,28 @@ public boolean hasIndex() {
...
+        if(!hasIndex()) {
             throw new SAMException("No index is available for this BAM file.");
+        }
         if(mIndex == null) {
-            SamIndexes samIndex = getIndexType();
-            if (samIndex == null) {
-                mIndex = mEnableIndexCaching ? new CachingBAMFileIndex(mIndexStream, getFileHeader().getSequenceDictionary())
-                        : new DiskBasedBAMFileIndex(mIndexStream, getFileHeader().getSequenceDictionary());
-            } else if (samIndex.equals(SamIndexes.BAI)) {
+            final SamIndexes samIndex = getIndexType();
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/BAMFileReader.java:414]: Can we rename samIndex -> samIndexType so it's more obvious it's an enum?
`SamIndexes` is not a super clear class name.

File: src/main/java/htsjdk/samtools/BAMFileReader.java [hunk 2]
```diff
@@ -407,20 +407,28 @@ public boolean hasIndex() {
...
+            if(mIndexFile != null) {
+                if (samIndex.equals(SamIndexes.BAI)) {
                     mIndex = mEnableIndexCaching ? new CachingBAMFileIndex(mIndexFile, getFileHeader().getSequenceDictionary(), mEnableIndexMemoryMapping)
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/BAMFileReader.java:417]: we might pull out a sequenceDictionary local variable to condense the 5 invocations of getFileHeader().getSequenceDictionary()

File: src/main/java/htsjdk/samtools/BAMFileReader.java [hunk 3]
```diff
@@ -438,8 +446,13 @@ public SamIndexes getIndexType() {
             } else if (mIndexFile.getName().toLowerCase().endsWith(FileExtensions.CSI)) {
                 return SamIndexes.CSI;
             }
-
             throw new SAMFormatException("Unknown BAM index file type: " + mIndexFile.getName());
+        } else if (mIndexStream != null) {
+            final SamIndexes samIndexes = SamIndexes.getSAMIndexTypeFromStream(mIndexStream);
```
Comment (lbergelson) [inline src/main/java/htsjdk/samtools/BAMFileReader.java:451]: same comment about the name here

**Review by lbergelson — CHANGES_REQUESTED** [review]
@cmnbroad I had a minor refactoring request for future clarity.  Do we still need the file paths at all?  Can we replace them with something opens a file stream instead?

File: src/main/java/htsjdk/samtools/BAMFileReader.java (hunk 1 above)
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/BAMFileReader.java:414]: Done. Would like to change the class name too. Maybe some day.

File: src/main/java/htsjdk/samtools/BAMFileReader.java (hunk 3 above)
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/BAMFileReader.java:451]: Done.

File: src/main/java/htsjdk/samtools/BAMFileReader.java (hunk 2 above)
Comment (cmnbroad) [reply src/main/java/htsjdk/samtools/BAMFileReader.java:417]: Done.

**Comment by cmnbroad** [comment]
Yeah the comments were all good improvements. Thanks for reviewing...

**Pull request merged**